File: 144-binary-tree-preorder-traversal/binary-tree-preorder-traversal.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
void solve(struct TreeNode* root, int* ans, int* returnSize) {
    if (root == NULL) return;

    ans[(*returnSize)++] = root->val;
    solve(root->left, ans, returnSize);
    solve(root->right, ans, returnSize);
}
int* preorderTraversal(struct TreeNode* root, int* returnSize) {
    *returnSize = 0;

    int MAX_NODE = 200;  // leetcode constrain
    int* ans = malloc(sizeof(struct TreeNode) * MAX_NODE);
    solve(root, ans, returnSize);

    return ans;
}

File: 144-binary-tree-preorder-traversal/binary-tree-preorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

def solve(root, ans: List[int]):
    if root == None: return

    ans.append(root.val)
    solve(root.left, ans)
    solve(root.right, ans)

class Solution:
    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        ans = []
        solve(root, ans)
        return ans